File: elixiraoc/lib/day4.ex
defmodule Day4 do
  def test do
    IO.puts("Train part1: #{part1("inputs/day4/train.txt")}")
    IO.puts("Test part1: #{part1("inputs/day4/test.txt")}")
    IO.puts("Train part2: #{part2("inputs/day4/train.txt")}")
    IO.puts("Test part2: #{part2("inputs/day4/test.txt")}")
  end

  def part1(path) do
    input(path)
    |> Enum.map(&String.split(&1, [",", "-"]))
    |> Enum.reduce(0, fn pairs, acc ->
      [first, second, third, fourth] = Enum.map(pairs, &String.to_integer/1)
      if ((first <= third) and (second >= fourth)) or (third <= first and fourth >= second) do
        acc + 1
      else
        acc
      end
    end)
  end

  def part2(path) do
    input(path)
    |> Enum.map(&String.split(&1, [",", "-"]))
    |> Enum.reduce(0, fn pairs, acc ->
      [first, second, third, fourth] = Enum.map(pairs, &String.to_integer/1)
      if ((second >= third) and (first <= fourth)) do
        acc + 1
      else
        acc
      end
    end)
  end

  def input(path), do: path |> File.stream!() |> Enum.map(&parse_input/1)

  def parse_input(input), do: input |> String.trim()
end

File: Python/day4/day4.py
from parse import parse

def part1(inputs):
    count = 0
    for input in inputs:
        if ((input[0] <= input[2]) and (input[1] >= input[3])) or ((input[2] <= input[0]) and (input[3] >= input[1])):
            count += 1

    return count

def part2(inputs):
    count = 0
    for input in inputs:
        if ((input[1] >= input[2]) and (input[0] <= input[3])):
            count += 1

    return count

if __name__ == "__main__":
    train = [parse("{:d}-{:d},{:d}-{:d}", line.strip()) for line in open("Python/day4/train.txt", "r")]
    test = [parse("{:d}-{:d},{:d}-{:d}", line.strip()) for line in open("Python/day4/test.txt", "r")]

    assert part1(train) == 2
    assert part1(test) == 536
    assert part2(train) == 4
    assert part2(test) == 845
    
    print(f"Train Part 1: {part1(train)}")
    print(f"Test Part 1: {part1(test)}")
    print(f"Train Part 2: {part2(train)}")
    print(f"Test Part 1: {part2(test)}")